modules/parachains/src/lib.rs

```diff
@@ -430,13 +522,30 @@ pub mod pallet {
 
 				let update_result: Result<_, ()> =
 					ParasInfo::<T, I>::try_mutate(parachain, |stored_best_head| {
+						let is_free = parachain_head_size <
+							T::ParaStoredHeaderDataBuilder::max_free_head_size() as usize &&
+							match stored_best_head {
+								Some(ref best_head)
+									if at_relay_block.0.saturating_sub(
+										best_head.best_head_hash.at_relay_block_number,
+									) >= free_headers_interval =>
+									true,
+								Some(_) => false,
+								None => true,
+							};
```

Comment:
Nit: I think part of this logic is duplicated here and in the call_ext. Maybe we can deduplicate.

Reply:
Same as here: https://github.com/paritytech/parity-bridges-common/pull/2873#discussion_r1560913279